pkg/services/ngalert/state/cache.go

```diff
@@ -78,6 +77,56 @@ func (c *cache) getOrCreate(alertRule *ngModels.AlertRule, result eval.Result) *
 	return newState
 }
 
+func (c *cache) expandRuleLabelsAndAnnotations(alertRule *ngModels.AlertRule, result eval.Result) (map[string]string, map[string]string) {
+	expand := func(original map[string]string) map[string]string {
+		expanded := make(map[string]string, len(original))
+		for k, v := range original {
+			ev, err := expandTemplate(alertRule.Title, v, result.Instance)
+			expanded[k] = ev
+			if err != nil {
+				c.log.Error("error in expanding template", "name", k, "value", v, "err", err.Error())
+				// Store the original template on error.
+				expanded[k] = v
+			}
+		}
+
+		return expanded
+	}
+
+	return expand(alertRule.Labels), expand(alertRule.Annotations)
+}
+
+func expandTemplate(name, text string, data map[string]string) (result string, resultErr error) {
+	name = "__alert_" + name
+	text = "{{$labels := .Labels}}" + text
```

Comment:
Wondering if this might create some extra whitespace? Been a bit since done templates, if so maybe `{{- ... -}}`?

Reply:
This wont create any additional whitespaces, but the additional `-` should be fine to add (will wait for any more comments before comitting this)

Reply:
Seems okay to leave then, up to you.